src/typechecker.c

```diff
@@ -899,24 +899,41 @@ NODISCARD bool typecheck_expression(AST *ast, Node *expr) {
 
       Type *t_from = expr->cast.value->type;
 
-      // TODO: What do we do when from and to types are equal?
-
-      if (type_is_incomplete(t_from))
-        ERR(expr->cast.value->source_location, "Can not cast from an incomplete type %T", t_from);
-
-      // FROM any integer type TO any integer type is ALLOWED
-      if (is_integer(t_from) && is_integer(t_to)) break;
-
-      // FROM any pointer type TO any integer type is ALLOWED
+      // TODO: Is complete to incomplete allowed?
+
+      // ALLOWED
+      // FROM any type T TO type T
+      if (types_equal(t_to, t_from)) break;
+      // FROM any pointer type TO any pointer type
+      // TODO: Check base type size + alignment...
+      if (is_pointer(t_from) && is_pointer(t_to)) break;
+      // FROM any pointer type TO any integer type
       if (is_pointer(t_from) && is_integer(t_to)) break;
+      // FROM any integer type TO any integer type
+      if (is_integer(t_from) && is_integer(t_to)) break;
 
-      // FROM any integer type TO any pointer type is DISALLOWED
+      // DISALLOWED
+      // FROM any integer type TO any pointer type is currently DISALLOWED, but very well may change
       if (is_integer(t_from) && is_pointer(t_to))
-        ERR(expr->source_location, "Can not cast from integer type %T to pointer type %T", t_from, t_to);
+        ERR(expr->cast.value->source_location,
+            "Can not cast from an integer type %T to pointer type %T",
+            t_from, t_to);
+
+      // FROM any incomplete type TO any complete type
+      if (type_is_incomplete(t_from) && !type_is_incomplete(t_to))
```

Comment:
I’d probably just disallow incomplete types in casting altogether, except for the case where you casting an incomplete type to itself, which should already be handled by the `types_equal()` check above (and casting *to* `void` should always be allowed as a special case to discard a value once/if we implement that that is)

Reply:
So `type_is_incomplete(t_from) || type_is_incomplete(t_to)` vs what we have now, right?

Reply:
Pretty much yeah. I’d put that check directly after the `types_equal()` check

Reply:
Implemented in https://github.com/LensPlaysGames/FUNCompiler/commit/db2d432d89f2c8e3e736b05dd91dd12b8c316cd6